### PR TITLE
feat(tsx): Prefix tsx output with a JSX pragma

### PR DESCRIPTION
**Pull request by Princesseuh**
## Changes

Add `/** @jsxImportSource astro */` to our TSX output so it uses the newly exported types at https://github.com/withastro/astro/pull/9501, this is all to attempt to fix https://github.com/withastro/language-tools/issues/727

## Testing

Adjusted tests

## Docs

N/A

### CODE REVIEW

**Comment by Princesseuh** [comment]
!preview tsx-prefix

File: internal/printer/print-to-tsx.go [hunk 1]
```diff
@@ -15,13 +15,17 @@ import (
...
+}
+
 func PrintToTSX(sourcetext string, n *Node, opts transform.TransformOptions, h *handler.Handler) PrintResult {
 	p := &printer{
 		sourcetext: sourcetext,
 		opts:       opts,
 		builder:    sourcemap.MakeChunkBuilder(nil, sourcemap.GenerateLineOffsetTables(sourcetext, len(strings.Split(sourcetext, "\n")))),
 	}
-
+	p.print(getTSXPrefix())
```
Comment (natemoo-re) [inline internal/printer/print-to-tsx.go:28]: If you move this inside of `renderTsx()` > `if n.Type == DocumentNode`, right after `hasChildren := false`, do you still get the weird behavior?

**Review by natemoo-re — COMMENTED** [review]
Looks fine to me!

File: internal/printer/print-to-tsx.go (hunk 1 above)
Comment (Princesseuh) [reply internal/printer/print-to-tsx.go:28]: Same behaviour, yeah. That's actually where I put it originally, ha!

File: internal/printer/print-to-tsx.go [hunk 2]
```diff
@@ -107,12 +111,12 @@ func renderTsx(p *printer, n *Node) {
...
 			if c.PrevSibling != nil && c.PrevSibling.Type == FrontmatterNode {
 				buf := strings.TrimSpace(string(p.output))
-				if len(buf) > 1 {
+				if len(buf)-len(getTSXPrefix()) > 1 {
 					char := rune(buf[len(buf)-1:][0])
 					// If the existing buffer ends with any character other than ;, we need to add a `;`
 					if char != ';' {
 						p.addNilSourceMapping()
-						p.print("\"\";")
+						p.print("{};")
```
Comment (Princesseuh) [inline internal/printer/print-to-tsx.go:119]: From my test, this seemed fine as a replacement of `"";`, but I feel like there must be a reason I went for `"";` back in the day. Can't remember it, though.

(When I first joined Astro, it was `;//`, from https://github.com/withastro/language-tools/pull/59 )

**Comment by Princesseuh** [comment]
!preview tsx-prefix

**Review by MoustaphaDev — APPROVED** [review]
Not sure why `""` was replaced by `{}` after frontmatter closing, but the code looks good to me!

**Comment by Princesseuh** [comment]
> The code looks good to me, but I wonder why `""` was replaced by `{}` after frontmatter closing?

Ah, I forgot to mention the justification, it's because it's seen as a string pragma (like use client, use strict) and so TypeScript will attempt to move it at the top of the file, which messes with imports

**Comment by MoustaphaDev** [comment]
Ah ok, makes sense now

**Pull request merged**